In [31]:
import sys, os
sys.path.append('/Users/raashidkhan/Desktop/GitHub-Projects/MNIST_Classification')
print(os.getcwd(), sys.path[0])
from src.training import load_split, get_or_train_model

/Users/raashidkhan/Desktop/GitHub-Projects/MNIST_Classification/notebook /opt/homebrew/Cellar/python@3.13/3.13.7/Frameworks/Python.framework/Versions/3.13/lib/python313.zip


In [32]:
X_train,y_train,X_test,y_test=load_split()
model = get_or_train_model()

In [33]:
y_train_predictions = model.predict(X_train)
import pandas as pd
pd.DataFrame({'Predicted': predictions, 'Actual': y_train==5})

Predicted  Actual
0          False   False
1          False   False
2          False   False
3          False   False
4          False   False
...          ...     ...
59995      False   False
59996       True    True
59997      False   False
59998      False   False
59999      False   False

[60000 rows x 2 columns]

In [34]:
print (predictions[10], y_train[10])

True 5


In [35]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_train==5, y_train_predictions)

array([[53338,  1242],
       [  922,  4498]])

True Negatives=53338
True Positives=4498
False Negative=922
False Positive=1242

Now let us say we reached perfection by:

In [36]:
perfect_predictions = y_train==5
confusion_matrix(y_train==5, perfect_predictions)

array([[54580,     0],
       [    0,  5420]])

So, there have been false negatives and false positives in our predictions.

Now, to know exact accuracy/performance of our model, we need to consider these things

precision = TP/TP + FP
recall = TP/TP + FN

In [37]:
from sklearn.metrics import precision_score, recall_score
print ('Precision Score: ',precision_score(y_train==5, y_train_predictions))
print ('Recall Score: ',recall_score(y_train==5, y_train_predictions))

Precision Score:  0.783623693379791
Recall Score:  0.8298892988929889


In [38]:
#Precision Score:  0.783623693379791
#Recall Score:  0.8298892988929889

Now, our model doesnt look as good as it was using cross_val_score
Now, we get why should we use confusion matrix to calculate performance of binary classifier